shards/watcher.go

```diff
@@ -237,9 +275,21 @@ func (s *DirectoryWatcher) watch() error {
 				}
 
 			case <-ticker.C:
-				// Periodically just double check the disk
+				// Periodically just double-check the disk
 				notify()
 
+				toAdd, toRemove := addOrRemove(watcher.WatchList(), tenant.ListIndexDirs(s.dir))
```

Comment:
I think this is pretty robust and simple. However, in the worst case we don't notice a new tenant for 1 minute. Once a tenant has been added to the watcher, new shards are picked up instantly.

Alternatively we could listen to events of `data/index/tenants` and add new tenants as they appear. I tried both versions and this one makes it more obvious that it reduces to the old code if tenancy is not enforced.

Reply:
alternatively as part of scan you can just always check the difference between WatchList and the dirs to watch? WatchList itself just holding a read-only mutex + creating in memory slice. ListIndexDirs has the same (or better) perf as the glob you do in scan.